roachtest: tpcc/mixed-headroom/n5cpu16 failed

SHA: https://github.com/cockroachdb/cockroach/commits/1051f376924f18c5782887ed824ab5ede5027493

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=tpcc/mixed-headroom/n5cpu16 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1429376&tab=buildLog

```
The test failed on branch=release-2.1, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20190808-1429376/tpcc/mixed-headroom/n5cpu16/run_1
	cluster.go:1735,tpcc.go:161,tpcc.go:163,tpcc.go:174,tpcc.go:287,test_runner.go:691: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1565244744-51-n5cpu16:5 -- ./workload fixtures load tpcc --warehouses=1000  {pgurl:1} returned:
		stderr:
		
		stdout:
		2.3 MiB)
		I190808 15:38:24.294129 48 ccl/workloadccl/fixture.go:544  loaded 887 KiB table district in 49.365343776s (10000 rows, 0 index entries, 18 KiB)
		I190808 15:46:13.672006 147 ccl/workloadccl/fixture.go:544  loaded 899 MiB table order in 8m38.743319862s (30000000 rows, 30000000 index entries, 1.7 MiB)
		I190808 15:54:04.953429 146 ccl/workloadccl/fixture.go:544  loaded 3.0 GiB table history in 16m30.02486995s (30000000 rows, 60000000 index entries, 3.1 MiB)
		I190808 16:05:15.465386 49 ccl/workloadccl/fixture.go:544  loaded 18 GiB table customer in 27m40.536546241s (30000000 rows, 30000000 index entries, 11 MiB)
		I190808 16:14:28.827232 150 ccl/workloadccl/fixture.go:544  loaded 32 GiB table stock in 36m53.898478143s (100000000 rows, 100000000 index entries, 15 MiB)
		I190808 16:26:22.588272 151 ccl/workloadccl/fixture.go:544  loaded 17 GiB table order_line in 48m47.659755157s (300011520 rows, 300011520 index entries, 5.8 MiB)
		Error: restoring fixture: pq: syntax error at or near "inject"
		Error:  exit status 1
		: exit status 1

```

Oh, regarding the failed splits -- this is unfortunately very likely just the whole CPut mess that wasn't fixed on the 19.x branch: https://github.com/cockroachdb/cockroach/pull/38302
And even if we did backport that, we don't require users to upgrade into the latest patch release.

The problem here is that once a descriptor has a 19.2-specific field, 19.1 nodes will always fail their cput:

```go

func TestFoo(t *testing.T) {
	s := []byte("\035\244\226 \003\010\205\023\022\t\300\211\367\001I\367\2205\210\032\010\300\211\367\001`\367\\\250\"\006\010\003\020\003\030\001\"\010\010\001\020\001\030\002 \000\"\006\010\004\020\004\030\003(\0040\025")
	v := Value{RawBytes: s}
	var desc RangeDescriptor
	require.NoError(t, v.GetProto(&desc))
	t.Error(pretty.Sprint(&desc))
}
```

the 19.1 node sees:

```
--- FAIL: TestFoo (0.00s)
    data_test.go:1522: &roachpb.RangeDescriptor{
            RangeID:  2437,
            StartKey: {0xc0, 0x89, 0xf7, 0x1, 0x49, 0xf7, 0x90, 0x35, 0x88},
            EndKey:   {0xc0, 0x89, 0xf7, 0x1, 0x60, 0xf7, 0x5c, 0xa8},
            Replicas: {
                {
                    NodeID:               3,
                    StoreID:              3,
                    ReplicaID:            1,
                    XXX_NoUnkeyedLiteral: struct {}{},
                    XXX_sizecache:        0,
                },
                {
                    NodeID:               1,
                    StoreID:              1,
                    ReplicaID:            2,
                    XXX_NoUnkeyedLiteral: struct {}{},
                    XXX_sizecache:        0,
                },
                {
                    NodeID:               4,
                    StoreID:              4,
                    ReplicaID:            3,
                    XXX_NoUnkeyedLiteral: struct {}{},
                    XXX_sizecache:        0,
                },
            },
            NextReplicaID:        4,
            Generation:           &int64(21),
            XXX_NoUnkeyedLiteral: struct {}{},
            XXX_sizecache:        0,
        }
FAIL

Process finished with exit code 1
```

but 19.2 reveals that the proto really has a replica with type `&zero` instead of `nil`.

```
  data_test.go:1891: &roachpb.RangeDescriptor{
            RangeID:          2437,
            StartKey:         {0xc0, 0x89, 0xf7, 0x1, 0x49, 0xf7, 0x90, 0x35, 0x88},
            EndKey:           {0xc0, 0x89, 0xf7, 0x1, 0x60, 0xf7, 0x5c, 0xa8},
            InternalReplicas: {
                {
                    NodeID:    3,
                    StoreID:   3,
                    ReplicaID: 1,
                    Type:      (*roachpb.ReplicaType)(nil),
                },
                {
                    NodeID:    1,
                    StoreID:   1,
                    ReplicaID: 2,
                    Type:      &roachpb.ReplicaType(0),
                },
                {
                    NodeID:    4,
                    StoreID:   4,
                    ReplicaID: 3,
                    Type:      (*roachpb.ReplicaType)(nil),
                },
            },
            NextReplicaID:        4,
            Generation:           &int64(21),
            GenerationComparable: (*bool)(nil),
            StickyBit:            (*hlc.Timestamp)(nil),
        }
```

On the plus side, the diff in my last post seems to have resolved the assertion crashes.

--- a/pkg/roachpb/metadata.go
+++ b/pkg/roachpb/metadata.go
@@ -145,7 +145,12 @@ func (r *RangeDescriptor) SetReplicaType(
 		desc := &r.InternalReplicas[i]
 		if desc.StoreID == storeID && desc.NodeID == nodeID {
 			prevTyp := desc.GetType()
-			desc.Type = &typ
+			if typ != VOTER_FULL {
+				desc.Type = &typ
+			} else {
+				// For 19.1 compatibility.
+				desc.Type = nil
+			}
 			return *desc, prevTyp, true
 		}
 	}
@@ -157,11 +162,16 @@ func (r *RangeDescriptor) SetReplicaType(
 func (r *RangeDescriptor) AddReplica(
 	nodeID NodeID, storeID StoreID, typ ReplicaType,
 ) ReplicaDescriptor {
+	var typPtr *ReplicaType
+	// For 19.1 compatibility, use nil instead of VOTER_FULL.
+	if typ != VOTER_FULL {
+		typPtr = &typ
+	}
 	toAdd := ReplicaDescriptor{
 		NodeID:    nodeID,
 		StoreID:   storeID,
 		ReplicaID: r.NextReplicaID,
-		Type:      &typ,
+		Type:      typPtr,
 	}
 	rs := r.Replicas()
 	rs.AddReplica(toAdd)

--- a/pkg/storage/batcheval/cmd_end_transaction.go
+++ b/pkg/storage/batcheval/cmd_end_transaction.go
@@ -1109,15 +1109,16 @@ func changeReplicasTrigger(
 
 	var desc roachpb.RangeDescriptor
 	if change.Desc != nil {
+		// Trigger proposed by a 19.2+ node (and we're a 19.2+ node as well).
 		desc = *change.Desc
 	} else {
+		// Trigger proposed by a 19.1 node. Reconstruct descriptor from deprecated
+		// fields.
 		desc = *rec.Desc()
 		desc.SetReplicas(roachpb.MakeReplicaDescriptors(change.DeprecatedUpdatedReplicas))
 		desc.NextReplicaID = change.DeprecatedNextReplicaID
 	}
 
-	// TODO(tschottdorf): duplication of Desc with the trigger below, should
-	// likely remove it from the trigger.
 	pd.Replicated.State = &storagepb.ReplicaState{
 		Desc: &desc,
 	}

--- a/pkg/storage/replica_command.go
+++ b/pkg/storage/replica_command.go
@@ -1266,9 +1266,12 @@ func (r *Replica) addReplicaLegacyPreemptiveSnapshot(
 	// complete. See #10409.
 	{
 		preemptiveRepDesc := roachpb.ReplicaDescriptor{
-			NodeID:    target.NodeID,
-			StoreID:   target.StoreID,
-			Type:      roachpb.ReplicaTypeVoterFull(),
+			NodeID:  target.NodeID,
+			StoreID: target.StoreID,
+			// NB: if we're still sending preemptive snapshot, the recipient is
+			// very likely a 19.1 node and does not understand this field. It
+			// won't matter to set it here, but don't anyway.
+			Type:      nil,
 			ReplicaID: 0, // intentional
 		}
 		if err := r.sendSnapshot(ctx, preemptiveRepDesc, SnapshotRequest_PREEMPTIVE, priority); err != nil {

--- a/pkg/storage/replica_command.go
+++ b/pkg/storage/replica_command.go
@@ -1406,9 +1406,10 @@ func execChangeReplicasTxn(
 			deprecatedRepDesc = removed[0]
 		}
 		crt = &roachpb.ChangeReplicasTrigger{
-			DeprecatedChangeType: deprecatedChangeType,
-			DeprecatedReplica:    deprecatedRepDesc,
-			Desc:                 &updatedDesc,
+			DeprecatedChangeType:      deprecatedChangeType,
+			DeprecatedReplica:         deprecatedRepDesc,
+			DeprecatedUpdatedReplicas: updatedDesc.Replicas().All(),
+			DeprecatedNextReplicaID:   updatedDesc.NextReplicaID,
 		}
 	} else {
 		crt = &roachpb.ChangeReplicasTrigger{

